Bluetooth: controller: Add terminate event for extended advertising

dagbja opened the pull request:
Add terminate event for duration timeout, max number of events and
connection.

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

dagbja committed:
Bluetooth: controller: Add terminate event for external advertising

Add terminate event for duration timeout, max number of events and
connection.

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

kruithofa commented on subsys/bluetooth/controller/include/ll.h:
```diff
@@ -53,7 +53,8 @@ uint8_t ll_adv_enable(uint8_t handle, uint8_t enable,
 		   uint8_t scan_window, uint8_t scan_delay);
 #else /* !CONFIG_BT_HCI_MESH_EXT */
-uint8_t ll_adv_enable(uint8_t handle, uint8_t enable);
+uint8_t ll_adv_enable(uint8_t handle, uint8_t enable,
```
Please fix the babblesim tests: 
the calls in tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c needs to be updated so that the simulator tests can run

dagbja replied:
Fixed

dagbja commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -222,16 +222,34 @@ static MFIFO_DEFINE(pdu_rx_free, sizeof(void *), PDU_RX_CNT);
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+#define LINK_RX_POOL_SIZE (sizeof(memq_link_t) * (RX_CNT + 2 + \
+						  BT_CTLR_MAX_CONN + \
+						  CONFIG_BT_CTLR_ADV_SET))
```
Yes, of course

dagbja replied:
Updated

dagbja replied:
Updated again to simpilfy the #define blocks

joerchan commented:
I tested this on nrf52840
```
west build tests/bluetooth/shell -- -DCONFIG_BT_CTLR_ADV_EXT=y
west flash --erase
```

Shell:
```
uart:~$ bt init
Bluetooth initialized
uart:~$ bt advertise on
Advertising started
Connected: 52:07:58:d6:cd:2e (random)
Remote LMP version 4.2 (0x08) subversion 0x6106 manufacturer 0x000f
LE Features: 0x00000000000000ff 
LE conn  param req: int (0x0006, 0x0006) lat 0 to 500
LE conn param updated: int 0x0006 lat 0 to 500
Disconnected: 52:07:58:d6:cd:2e (random) (reason 0x08)
```
This happens everytime, does not happen on the commit preceding this PR.

dagbja commented:
Fixed the disconnect issue (I hope!)

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3564,6 +3565,37 @@ static void le_adv_ext_coded_report(struct pdu_data *pdu_data,
+
+	/* Bit 17 set indicates event was caused by connection */
+	/* Bit 16 set indicates event was caused by duration timeout */
+	if ((uint32_t)node_rx->hdr.rx_ftr.extra & BIT(TERM_EVT_CONNECT)) {
```
Return the status in bits 16-23 directly instead of using bit fields?

dagbja replied:
Where do I send status? node_rx_hdr has extra and param fields, which we use for # of elapsed events and conn handle. Since we only use 16 out of 32 bits in extra, that seemed like a good place for these bits, which determine status. But we could also put status (8 bits) directly in bits 16-23 of extra.

dagbja replied:
Added status as bit 8-15 in extra.

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -39,6 +39,11 @@
 #endif /* !CONFIG_BT_CTLR_ADV_EXT */
 #endif /* CONFIG_BT_OBSERVER */
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
```
if you return status, you will not need these?

dagbja replied:
Removed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1722,13 +1747,70 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+	hdr = &adv->ull;
+	mfy.param = &adv->lll;
+
+	if (hdr->ref) {
```
will this always be true by design? if so, is this check required?

dagbja replied:
Removed this function. I copied this functionality from ull_adc.c (ticker_op_stop_cb). I assume we don't need to call lll_disable, as we do in ull_adv.c when ref count is 0. lll_disable probably gets called as a result of an HCI call?

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1722,13 +1747,70 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+	if (hdr->ref) {
+		LL_ASSERT(!hdr->disabled_cb);
+		hdr->disabled_param = NULL;
+		hdr->disabled_cb = NULL;
```
As the assert above checks for being NULL, these assignment is not needed.

dagbja replied:
It's done in ull_adv.c (ticker_op_stop_cb). Should we remove them there too? Or keep both?

dagbja replied:
Actually, this is the function I removed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1722,13 +1747,70 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+		hdr->disabled_param = NULL;
+		hdr->disabled_cb = NULL;
+
+		mfy.fp = lll_disable;
```
How is the race condition handled with advertising disabled by HCI command?

dagbja replied:
Removed this function.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1722,13 +1747,70 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+	adv = (void *)HDR_LLL2EVT(lll);
+
+	handle = ull_adv_handle_get(adv);
+	LL_ASSERT(handle < BT_CTLR_ADV_SET);
```
This check is not needed as long as ull_hdr supplied by caller is valid, instead add checks in caller if needed (helps in debugging caller).

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1722,13 +1747,70 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
 	struct lll_event *next;
 	void *release;
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	struct lll_adv *lll;
```
Move these close to locality of reference, inside the `case` .

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1741,6 +1823,49 @@ static inline void rx_demux_event_done(memq_link_t *link,
+
+			lll->node_rx_adv_term->rx_ftr.extra = (void *)
+				((uint32_t)adv->event_counter
+					| BIT(TERM_EVT_TIMEOUT));
```
pass HCI error code instead.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -820,10 +821,44 @@ uint8_t ll_adv_enable(uint8_t enable)
 	}
+
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	if (adv->is_created & BIT(1)) {
```
use already defined defines instead of BIT(1).

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1104,6 +1139,10 @@ uint8_t ll_adv_enable(uint8_t enable)
 #endif /* !CONFIG_BT_HCI_MESH_EXT */
 #endif /* CONFIG_BT_CTLR_PRIVACY */
 
+	lll_rand_isr_get(&adv->next_random_delay, sizeof(adv->next_random_delay));
```
this is not an ISR context.

dagbja replied:
Replaced the call with assignment to 0.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_types.h:
```diff
@@ -19,10 +19,15 @@ struct ll_adv_set {
 	uint16_t interval;
 #endif /* !CONFIG_BT_CTLR_ADV_EXT */
 
+	uint32_t next_random_delay;
```
I do not see the need for this. Additionally you are adding overhead to legacy advertising feature and could complicate if conditional compiling extended advertising.

dagbja replied:
I don't see another solution Asked Peter Hansen.
Keeping the code until we have another solution.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -214,6 +214,28 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
 	}
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	if (adv->is_created & BIT(1)) {
```
use define

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -820,10 +821,44 @@ uint8_t ll_adv_enable(uint8_t enable)
+	ull = &adv->ull;
+	adv->max_events = max_ext_adv_evts;
+	adv->event_counter = 0;
+	adv->remain_duration = HAL_TICKER_US_TO_TICKS(duration*10000);
```
Store in units of 625us.

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1387,29 +1432,51 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
+
+	adv->event_counter += (lazy + 1);
+
+	interval = HAL_TICKER_US_TO_TICKS
+		((uint64_t)adv->interval * 625U);
+
+	ticks_next_evt_done = 2*interval
+		+ curr_random_delay + adv->next_random_delay;
+
+	if (adv->remain_duration) {
+		if (adv->remain_duration >= ticks_next_evt_done) {
+			adv->remain_duration -= ticks_next_evt_done;
+		} else {
+			/* A value of 1 means send term event */
+			adv->remain_duration = 1;
+		}
+	}
```
Simply these based on use of 625us units. Using a value of 1 is unnecessary, a check `remain <= interval` in the event done would suffice, if `remain` has already been pre-deducted of the random_delay.

dagbja replied:
The value 1 means that rx_demux_event_done will disable set, then set this to 0. If rx_demux_event_done used 0 as a criteria, it would happen when timeout is not enabled from host (duration=0).

joerchan commented:
In the current version of this PR the conn_handle and number of adv events appear to not be valid.

```
[00:02:08.827,911] <wrn> bt_hci_core: status 0x00 handle 0 role 1 peer 40:1a:37:15:9a:d0 (random) peer RPA 00:00:00:00:00:00
[00:02:08.828,002] <wrn> bt_hci_core: local RPA 00:00:00:00:00:00
[00:02:08.833,648] <wrn> bt_hci_core: status 0x00 adv_handle 0 conn_handle 0x6da0 num 142
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -515,7 +533,7 @@ uint8_t ll_rx_get(void **node_rx, uint16_t *handle)
 			if (rx->type == NODE_RX_TYPE_DC_PDU_RELEASE) {
 				(void)memq_dequeue(memq_ll_rx.tail,
-						   &memq_ll_rx.head, NULL);
+					&memq_ll_rx.head, NULL);
```
revert indentation changes.

dagbja replied:
Fixed. I though leading spaces was not allowed, that's why I changed it.  (I guess spaces that can not be replaced with tab are allowed, i.e., less than 8 spaces.)
Fixed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -550,7 +568,7 @@ void ll_rx_dequeue(void)
 
 	link = memq_dequeue(memq_ll_rx.tail, &memq_ll_rx.head,
-			    (void **)&rx);
+		(void **)&rx);
```
revert indentation changes.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -646,9 +664,9 @@ void ll_rx_dequeue(void)
-			      ull_scan_is_enabled(0) << 1) |
-			     (IS_ENABLED(CONFIG_BT_BROADCASTER) &&
-			      ull_adv_is_enabled(0));
+				ull_scan_is_enabled(0) << 1) |
```
revert indentation changes.

dagbja replied:
(See previous comment.)

cvinayak replied:
did you miss this one? doesn't seem to have been reverted

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -889,7 +913,7 @@ void ll_rx_mem_release(void **node_rx)
 			LL_ASSERT(!conn->lll.link_tx_free);
 			link = memq_deinit(&conn->lll.memq_tx.head,
-					   &conn->lll.memq_tx.tail);
+				&conn->lll.memq_tx.tail);
```
revert indentation changes.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -55,13 +55,13 @@ inline uint16_t ull_adv_handle_get(struct ll_adv_set *adv);
-static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t lazy,
-		      void *param);
+static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
+	uint16_t lazy, void *param);
```
Use leading tabs, and then spaces to align with parameters.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -55,13 +55,13 @@ inline uint16_t ull_adv_handle_get(struct ll_adv_set *adv);
-static void ticker_stop_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t lazy,
-			   void *param);
+static void ticker_stop_cb(uint32_t ticks_at_expire, uint32_t remainder,
+	uint16_t lazy, void *param);
```
Use leading tabs, and then spaces to align with parameters.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1421,8 +1496,8 @@ static void ticker_op_update_cb(uint32_t status, void *param)
-static void ticker_stop_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t lazy,
-			   void *param)
+static void ticker_stop_cb(uint32_t ticks_at_expire, uint32_t remainder,
+	uint16_t lazy, void *param)
```
Use leading tabs, and then spaces to align with parameters.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -348,8 +369,8 @@ void ull_slave_done(struct node_rx_event_done *done, uint32_t *ticks_drift_plus,
-void ull_slave_ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t lazy,
-			 void *param)
+void ull_slave_ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
+	uint16_t lazy, void *param)
```
Use leading tabs, and then spaces to align with parameters.

dagbja replied:
(See previous comment.)

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -820,10 +821,45 @@ uint8_t ll_adv_enable(uint8_t enable)
+		struct node_rx_pdu *node_rx_adv_term;
+		void *link_adv_term;
+
+		/* The alloc here used for connection complete event */
```
is this comment correct?

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -820,10 +821,45 @@ uint8_t ll_adv_enable(uint8_t enable)
+		/* The alloc here used for connection complete event */
+		link_adv_term = ll_rx_link_alloc();
+		if (!link_adv_term) {
+				/* TODO: figure out right return value */
```
indentation of the comment

dagbja replied:
Fixed

cvinayak committed:
Bluetooth: controller: Adv Set Terminated Evt on duration

Implementation of Advertising Set Terminated Event on
duration sans use of additional variable in advertising
set context structure.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

dagbja committed:
Bluetooth: controller: Terminate event for ext adv

Update to free term event node

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1602,8 +1602,12 @@ static inline uint8_t disable(uint8_t handle)
 	void *mark;
 	uint32_t ret;
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	struct node_rx_pdu *node_rx_adv_term;
```
move this inside the if clause

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1669,6 +1673,16 @@ static inline uint8_t disable(uint8_t handle)
 #endif /* CONFIG_BT_PERIPHERAL */
 
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	lll = &adv->lll;
```
```suggestion
	struct lll_adv *lll = &adv->lll;
```

cvinayak replied:
and remove declaration from line 1605

dagbja replied:
Fixed

dagbja committed:
Bluetooth: controller: Release term event memory on timeout

Release memory allocated to ext adv terminate event, after timeout

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

carlescufi commented:
@cvinayak @kruithofa could you please re-review?

wopu-ot commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3564,6 +3565,26 @@ static void le_adv_ext_coded_report(struct pdu_data *pdu_data,
+
+	sep = meta_evt(buf, BT_HCI_EVT_LE_ADV_SET_TERMINATED, sizeof(*sep));
+	sep->num_completed_ext_adv_evts =
+		(uint8_t)((uint32_t)node_rx->hdr.rx_ftr.extra & 0xff);
```
I would prefer if rx_ftr would have explicit fields to carry the respective information instead of using the generic "extra" field. If space is of the essence, use a union.

cvinayak committed:
tests: Bluetooth: bsim: Added tests for adv set terminate event

Added bsim testing for Advertising Set Terminated event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Reworked adv set terminate event

Reworked the implementation of Advertising Set Terminate
event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@dagbja I have now reworked, and testing using bsim. Please review and test the changes.

dagbja commented:
OK. Thanks!

From: Vinayak Kariappa Chettimada <notifications@github.com>
Sent: Monday, July 13, 2020 2:55 PM
To: zephyrproject-rtos/zephyr <zephyr@noreply.github.com>
Cc: Bjarvin, Dag <Dag.Bjarvin@nordicsemi.no>; Mention <mention@noreply.github.com>
Subject: Re: [zephyrproject-rtos/zephyr] Bluetooth: controller: Add terminate event for extended advertising (#26338)


@dagbja<https://github.com/dagbja> I have now reworked, and testing using bsim. Please review and test the changes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/zephyrproject-rtos/zephyr/pull/26338#issuecomment-657543436>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOPL5BV5PWNUQV26MGBH33LR3L73XANCNFSM4OESUCFQ>.

joerchan commented:
Tested and now it works as expected.

cvinayak commented:
> 
> 
> Tested and now it works as expected.

@joerchan Thank you. Will rework review comments from @wopu-ot and then should be good to go in.

cvinayak committed:
Bluetooth: controller: Define struct in node rx for adv term event

Defined struct in node rx for passing event parameters of
Advertising Set Terminated Event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Added some comments for fields in node rx struct

Added internal comments about the use of some of the fields
in the node rx structure.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@wopu-ot please re-review

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	printk("success.\n");
+
+	printk("Waiting...");
+	while (is_sent) {
```
is_sent is initialised to false here, and set to true in the sent_cb. As a result the sleep loop in line 221 is never executed (if it was it would be a forever-loop, never continuing). 
If we want to make sure that advertisement has started it should be **while (!is_sent)** and a timeout mechanism added
If the intention is to wait here until the advertising is terminated we need to set is_sent to false in the appropriate place, and we still need to add the code to ensure that advertisement has started

cvinayak replied:
Fixed

kruithofa replied:
Ok, note however this test still assumes that the function is working correct, otherwise you'll have an endless loop. I.e. it is good enough for debugging, but not for an automated test like C.I.

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	while (is_sent) {
+		k_sleep(K_MSEC(100));
+	}
+	printk("done.\n");
```
Even though we say that we are done, we are not testing that advertisement has stopped after the correct number of events. 
in other words: when we come here it can be for several reasons that should cause the test to fail
- advertisement never started
- advertisement has started, but is not stopped
- advertisement has started, and is stopped, but after the wrong number of events

cvinayak replied:
Check for num_sent added on in the advx test. Test to check for adv reports in scanx test is todo, postponed to be added as part of periodic advertising testing.

kruithofa replied:
Ok, looking forward to this in the next PR

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	}
+	printk("done.\n");
+
+	k_sleep(K_MSEC(1000));
```
This sleep should not be necessary; currently we need it because the previous test is not complete.
Please fix the previous test, and remove the sleep

cvinayak replied:
The sleep is added to have clear differentiation (gaps) when analyzing exported Ellisys traces.

kruithofa replied:
A very good reason. Nevertheless if this should be part of C.I. testing the delays should be removed. 
Also note that the sleep at line 430 is suspect; it will make the tests fail if the values set here is too large

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+
+	k_sleep(K_MSEC(1000));
+
+	printk("Start advertising using extended commands (duration)...");
```
Same comments as for the previous test

cvinayak replied:
The sleep is added to have clear differentiation (gaps) when analyzing exported Ellisys traces.

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	if (err) {
+		goto exit;
+	}
+	printk("success.\n");
```
We print success based on the result of the HCI command, but we need to check that the LL is actually sending out advertisement packets

cvinayak replied:
Check for num_sent added on in the advx test. Test to check for adv reports in scanx test is todo, postponed to be added as part of periodic advertising testing.

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	printk("success.\n");
+
+	printk("Waiting...");
+	while (is_sent) {
```
see the comment for line 220

cvinayak replied:
Fixed

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	}
+	printk("success.\n");
+
+	k_sleep(K_MSEC(400));
```
We need to verify that advertising indeed has started, and then this sleep is not necessary

cvinayak replied:
The sleep is added to have clear differentiation (gaps) when analyzing exported Ellisys traces.

kruithofa commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -105,6 +199,70 @@ static void test_advx_main(void)
+	}
+	printk("success.\n");
+
+	k_sleep(K_MSEC(1000));
```
We need to verify that advertising indeed has stopped, and then this sleep is not necessary

cvinayak replied:
The sleep is added to have clear differentiation (gaps) when analyzing exported Ellisys traces.

kruithofa commented:
Approved, but note comments
